mgmtsystem/hawkular.py

```diff
@@ -766,6 +788,372 @@ def delete_resource(self, feed_id, resource_id):
+        """Returns list of datasources (both XA and non XA).
+
+         Args:
+             feed_id: Feed id of the datasource (optional)
+        """
+        resources = self.list_resource(feed_id=feed_id,
+                                       resource_type_id='Datasource',
+                                       list_children=True)
+        resources.extend(self.list_resource(
+            feed_id=feed_id,
+            resource_type_id='XA Datasource', list_children=True))
+        datasources = []
+        if resources:
+            for resource in resources:
+                datasources.append(Datasource(resource.id, resource.name, resource.path))
+        return datasources
+
+    def list_resource(self, resource_type_id, feed_id=None,
+                      list_children=False, include_data=False):
+        """Returns list of resources.
+
+          Args:
+            feed_id: Feed id of the resource (optional)
+            resource_type_id: Resource type id
+            list_children: whether recursively list child resources (optional)
+            include_data: whether to include data value of resource (optional)
+        """
+        if not feed_id:
+            resources = []
+            for feed in self.list_feed():
+                resources.extend(self._list_resource(feed_id=feed.path.feed_id,
+                                                     resource_type_id=resource_type_id,
+                                                     list_children=list_children,
+                                                     include_data=include_data))
+            return resources
+        else:
+            return self._list_resource(feed_id=feed_id, resource_type_id=resource_type_id,
+                                       list_children=list_children,
+                                       include_data=include_data)
+
+    def _list_resource(self, feed_id, resource_type_id=None,
+                       list_children=False, include_data=False):
+        """Returns list of resources.
+
+         Args:
+            feed_id: Feed id of the resource
+            resource_type_id: Resource type id (optional)
+            list_children: whether recursively list child resources (optional)
+            include_data: whether to include data value of resource (optional)
+        """
+        if not feed_id:
+            raise KeyError("'feed_id' is a mandatory field!")
+        entities = []
+        if resource_type_id:
+            result = self._post_raw('strings/raw/query',
+                data={"fromEarliest": "true", "order": "DESC",
+                    "tags": "feed:{},type:r,restypes:.*\\|{}\\|.*".
+                    format(feed_id, resource_type_id)})
+        else:
+            result = self._post_raw('strings/raw/query',
+                data={"fromEarliest": "true", "order": "DESC",
+                    "tags": "feed:{},type:r".format(feed_id)})
+
+        if result.status_code == 200:
```

Comment:
To avoid more complexity, how about shortcutting it:
```python
if result.status_code != 200:
    return entities
for entity_j.....
```

Allows us to dedent once at least ;)